tests/unit/s2n_ktls_cmsg_test.c

```diff
@@ -0,0 +1,365 @@
+#include "testlib/s2n_testlib.h"
+#include "tls/s2n_ktls.h"
+#include "utils/s2n_random.h"
+
+S2N_RESULT s2n_ktls_send_msg_impl(int sock, struct msghdr *msg,
+        struct iovec *msg_iov, size_t count, s2n_blocked_status *blocked, ssize_t *send_len);
+S2N_RESULT s2n_ktls_recv_msg_impl(struct s2n_connection *conn, int sock, struct msghdr *msg,
+        struct iovec *msg_iov, s2n_blocked_status *blocked, ssize_t *bytes_read);
+S2N_RESULT s2n_ktls_set_ancillary_data(struct msghdr *msg, uint8_t record_type);
+S2N_RESULT s2n_ktls_parse_ancillary_data(struct msghdr *msg, uint8_t *record_type);
+
+#define TEST_MAX_DATA_LEN 20000
+uint8_t TEST_SEND_RECORD_TYPE = 10;
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+#if !S2N_KTLS_SUPPORTED /* CMSG_* macros are platform specific */
+    char buf[sizeof(uint8_t)] = { 0 };
+
+    /* Init msghdr */
+    struct msghdr msg = { 0 };
+    msg.msg_control = buf;
+    msg.msg_controllen = sizeof(uint8_t);
+
+    EXPECT_ERROR_WITH_ERRNO(s2n_ktls_set_ancillary_data(&msg, TEST_SEND_RECORD_TYPE), S2N_ERR_KTLS_UNSUPPORTED_PLATFORM);
+    uint8_t record_type = 0;
+    EXPECT_ERROR_WITH_ERRNO(s2n_ktls_parse_ancillary_data(&msg, &record_type), S2N_ERR_KTLS_UNSUPPORTED_PLATFORM);
+
+#else /* kTLS supported */
+
+    uint8_t test_data[TEST_MAX_DATA_LEN] = { 0 };
+    struct s2n_blob test_data_blob = { 0 };
+    EXPECT_SUCCESS(s2n_blob_init(&test_data_blob, test_data, sizeof(test_data)));
+    EXPECT_OK(s2n_get_public_random_data(&test_data_blob));
+
+    /* Test send/recv msg */
+    {
+        /* ctrl_msg send and recv data */
+        for (size_t to_send = 1; to_send < TEST_MAX_DATA_LEN; to_send += 500) {
+            DEFER_CLEANUP(struct s2n_connection *conn = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            DEFER_CLEANUP(struct s2n_test_io_pair io_pair = { 0 }, s2n_io_pair_close);
+            EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+            s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+
+            /* init send msg */
+            struct msghdr send_msg = { 0 };
+            struct iovec send_msg_iov = { 0 };
+            send_msg_iov.iov_base = (void *) (uintptr_t) test_data;
+            send_msg_iov.iov_len = to_send;
+            /* init rev msg */
+            uint8_t recv_buffer[TEST_MAX_DATA_LEN] = { 0 };
+            struct msghdr recv_msg = { 0 };
+            struct iovec recv_msg_iov = { 0 };
+            recv_msg_iov.iov_base = recv_buffer;
+            recv_msg_iov.iov_len = to_send;
+
+            /* send msg */
+            ssize_t sent_len = 0;
+            EXPECT_OK(s2n_ktls_send_msg_impl(io_pair.client, &send_msg, &send_msg_iov, 1, &blocked, &sent_len));
+            EXPECT_EQUAL(sent_len, to_send);
+
+            /* recv msg */
+            ssize_t recv_len = 0;
+            EXPECT_OK(s2n_ktls_recv_msg_impl(conn, io_pair.server, &recv_msg, &recv_msg_iov, &blocked, &recv_len));
+            EXPECT_EQUAL(recv_len, to_send);
```

Comment:
I'm not sure you can guarantee that either of these won't partially complete. Like if the tx socket has a small buffer or the TCP decides to split the recv into multiple chunks

Reply:
Sure, but I dont think we will hit that case in CI testing.

The partial reads/send will be handled by the kTLS send/recv fn. I will also mock out out the IO when I add that functionality and test for that.

Reply:
I'm saying you _could_ hit that case in CI or on someone's random machine, making this a flaky test.

Reply:
addressed in https://github.com/aws/s2n-tls/pull/4073/commits/3deb30252abe07f8ab07ccc8e36fe0532c9674b4

yea I guess its a bit more likely since we are sending larger data in a loop.